File: clientflow-ai/sdk/python/clientflow_sdk.py
"""
PROMPT 117 — ClientFlow AI Python SDK
A lightweight API client for the ClientFlow AI REST API.

Installation:
    pip install requests
    # then copy this file or install when published to PyPI

Usage:
    from clientflow_sdk import ClientFlow

    cf = ClientFlow(base_url='https://api.clientflow.ai/api/v1')
    cf.auth.login('admin@acme.com', 'password', org_slug='acme')

    clients = cf.clients.list(status='active')
    for c in clients['clients']:
        print(c['name'])
"""

import hashlib
import json
from typing import Any, Dict, Optional
import urllib.request
import urllib.parse
import urllib.error


class ClientFlowError(Exception):
    """Raised when the API returns an error response."""
    def __init__(self, message: str, status: int, code: str = None, details: Any = None):
        super().__init__(message)
        self.status  = status
        self.code    = code
        self.details = details


class ClientFlowSDK:
    """
    ClientFlow AI API client.

    :param base_url: API base URL, e.g. 'https://api.clientflow.ai/api/v1'
    :param token:    Optional JWT token (can be set after login)
    :param timeout:  Request timeout in seconds (default: 30)
    """

    def __init__(
        self,
        base_url: str = "https://api.clientflow.ai/api/v1",
        token: Optional[str] = None,
        timeout: int = 30,
    ):
        self.base_url = base_url.rstrip("/")
        self.token    = token
        self.timeout  = timeout

        # Resource namespaces
        self.auth        = AuthResource(self)
        self.clients     = ClientsResource(self)
        self.broadcasts  = BroadcastsResource(self)
        self.payments    = PaymentsResource(self)
        self.analytics   = AnalyticsResource(self)
        self.ai          = AIResource(self)
        self.sessions    = SessionsResource(self)
        self.experiments = ExperimentsResource(self)
        self.branding    = BrandingResource(self)
        self.usage       = UsageResource(self)

    def set_token(self, token: str) -> None:
        """Set JWT token (called automatically after login)."""
        self.token = token

    def _request(
        self,
        method: str,
        path: str,
        body: Optional[Dict] = None,
        params: Optional[Dict] = None,
    ) -> Any:
        url = self.base_url + path
        if params:
            filtered = {k: v for k, v in params.items() if v is not None}
            if filtered:
                url += "?" + urllib.parse.urlencode(filtered)

        data = json.dumps(body).encode("utf-8") if body is not None else None

        headers = {"Content-Type": "application/json", "Accept": "application/json"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"

        req = urllib.request.Request(url, data=data, headers=headers, method=method)

        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                raw = resp.read()
                return json.loads(raw) if raw else {}
        except urllib.error.HTTPError as e:
            raw = e.read()
            try:
                payload = json.loads(raw)
            except Exception:
                payload = {}
            raise ClientFlowError(
                message=payload.get("error", f"HTTP {e.code}"),
                status=e.code,
                code=payload.get("code"),
                details=payload.get("details"),
            ) from e
        except urllib.error.URLError as e:
            raise ClientFlowError(message=str(e.reason), status=0) from e

    def _get(self, path: str, params: Optional[Dict] = None) -> Any:
        return self._request("GET", path, params=params)

    def _post(self, path: str, body: Optional[Dict] = None) -> Any:
        return self._request("POST", path, body=body)

    def _put(self, path: str, body: Optional[Dict] = None) -> Any:
        return self._request("PUT", path, body=body)

    def _delete(self, path: str) -> Any:
        return self._request("DELETE", path)


# ─── Resources ────────────────────────────────────────────────────────────────

class AuthResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def login(self, email: str, password: str, org_slug: Optional[str] = None) -> Dict:
        """Login and automatically set the token."""
        body: Dict = {"email": email, "password": password}
        if org_slug:
            body["orgSlug"] = org_slug
        data = self._sdk._post("/auth/login", body)
        if "token" in data:
            self._sdk.set_token(data["token"])
        return data

    def register_org(self, org_name: str, org_slug: str, email: str, password: str, name: str = None) -> Dict:
        return self._sdk._post("/auth/register-org", {
            "orgName": org_name, "orgSlug": org_slug,
            "email": email, "password": password, "name": name,
        })

    def me(self) -> Dict:
        return self._sdk._get("/auth/me")

    def logout(self) -> Dict:
        result = self._sdk._post("/auth/logout")
        self._sdk.token = None
        return result


class ClientsResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def list(self, status: str = None, search: str = None, page: int = 1, limit: int = 50) -> Dict:
        return self._sdk._get("/clients", {"status": status, "search": search, "page": page, "limit": limit})

    def get(self, client_id: str) -> Dict:
        return self._sdk._get(f"/clients/{client_id}")

    def create(self, data: Dict) -> Dict:
        return self._sdk._post("/clients", data)

    def update(self, client_id: str, data: Dict) -> Dict:
        return self._sdk._put(f"/clients/{client_id}", data)

    def delete(self, client_id: str) -> Dict:
        return self._sdk._delete(f"/clients/{client_id}")


class BroadcastsResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def list(self) -> Dict:
        return self._sdk._get("/broadcasts")

    def get(self, broadcast_id: str) -> Dict:
        return self._sdk._get(f"/broadcasts/{broadcast_id}")

    def create(self, data: Dict) -> Dict:
        return self._sdk._post("/broadcasts", data)

    def send(self, broadcast_id: str) -> Dict:
        return self._sdk._post(f"/broadcasts/{broadcast_id}/send")


class PaymentsResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def list(self, status: str = None) -> Dict:
        return self._sdk._get("/payments", {"status": status})

    def get(self, payment_id: str) -> Dict:
        return self._sdk._get(f"/payments/{payment_id}")

    def create(self, data: Dict) -> Dict:
        return self._sdk._post("/payments", data)

    def confirm(self, payment_id: str) -> Dict:
        return self._sdk._post(f"/payments/{payment_id}/confirm")


class AnalyticsResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def dashboard(self) -> Dict:
        return self._sdk._get("/analytics")

    def revenue(self) -> Dict:
        return self._sdk._get("/analytics/revenue")

    def ai_usage(self) -> Dict:
        return self._sdk._get("/analytics/ai")


class AIResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def reply(self, client_id: str, message: str) -> Dict:
        return self._sdk._post("/ai/reply", {"clientId": client_id, "message": message})

    def suggest(self, client_id: str, context: str) -> Dict:
        return self._sdk._post("/ai/suggest", {"clientId": client_id, "context": context})


class SessionsResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def list(self) -> Dict:
        return self._sdk._get("/sessions")

    def revoke(self, session_id: str) -> Dict:
        return self._sdk._delete(f"/sessions/{session_id}")

    def revoke_all(self) -> Dict:
        return self._sdk._delete("/sessions")


class ExperimentsResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def list(self) -> Dict:
        return self._sdk._get("/experiments")

    def create(self, data: Dict) -> Dict:
        return self._sdk._post("/experiments", data)

    def start(self, experiment_id: str) -> Dict:
        return self._sdk._put(f"/experiments/{experiment_id}/start")

    def results(self, experiment_id: str) -> Dict:
        return self._sdk._get(f"/experiments/{experiment_id}/results")

    def set_winner(self, experiment_id: str, winner: str) -> Dict:
        return self._sdk._post(f"/experiments/{experiment_id}/winner", {"winner": winner})

    def record_conversion(self, experiment_id: str, client_id: str) -> Dict:
        return self._sdk._post(f"/experiments/{experiment_id}/convert", {"clientId": client_id})


class BrandingResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def get(self) -> Dict:
        return self._sdk._get("/branding")

    def update(self, data: Dict) -> Dict:
        return self._sdk._put("/branding", data)

    def init_domain(self, domain: str) -> Dict:
        return self._sdk._post("/domains/initiate", {"domain": domain})

    def verify_domain(self) -> Dict:
        return self._sdk._post("/domains/verify")

    def get_domain(self) -> Dict:
        return self._sdk._get("/domains")


class UsageResource:
    def __init__(self, sdk: ClientFlowSDK):
        self._sdk = sdk

    def monthly(self) -> Dict:
        return self._sdk._get("/usage/monthly")

    def top_features(self) -> Dict:
        return self._sdk._get("/usage/top-features")


# ─── Convenience alias ────────────────────────────────────────────────────────
ClientFlow = ClientFlowSDK


# ─── Example usage ────────────────────────────────────────────────────────────
if __name__ == "__main__":
    cf = ClientFlow(base_url="http://localhost:5000/api/v1")

    # Login
    auth_data = cf.auth.login("admin@acme.com", "secret123", org_slug="acme")
    print("Logged in as:", auth_data.get("email"))

    # List clients
    result = cf.clients.list(status="active", limit=10)
    print("Active clients:", len(result.get("clients", [])))

    # Analytics
    stats = cf.analytics.dashboard()
    print("Dashboard stats:", stats)

    # Logout
    cf.auth.logout()
    print("Logged out")

File: clientflow-ai/sdk/js/clientflow-sdk.js
/**
 * PROMPT 117 — ClientFlow AI JavaScript SDK
 * A lightweight API client for the ClientFlow AI REST API.
 *
 * Installation:
 *   npm install clientflow-ai-sdk
 *   // or copy this file directly
 *
 * Usage:
 *   const ClientFlow = require('clientflow-ai-sdk');
 *   const cf = new ClientFlow({ baseUrl: 'https://api.clientflow.ai', token: 'your-jwt' });
 *   const clients = await cf.clients.list();
 */

class ClientFlowSDK {
  /**
   * @param {Object} config
   * @param {string} config.baseUrl   API base URL (e.g. 'https://api.clientflow.ai/api/v1')
   * @param {string} [config.token]   JWT token (set after login, or pass directly)
   * @param {number} [config.timeout] Request timeout in ms (default: 30000)
   */
  constructor({ baseUrl, token, timeout = 30000 } = {}) {
    this.baseUrl = (baseUrl || 'http://localhost:5000/api/v1').replace(/\/$/, '');
    this.token   = token || null;
    this.timeout = timeout;

    // Resource namespaces
    this.auth        = new AuthResource(this);
    this.clients     = new ClientsResource(this);
    this.broadcasts  = new BroadcastsResource(this);
    this.payments    = new PaymentsResource(this);
    this.analytics   = new AnalyticsResource(this);
    this.ai          = new AIResource(this);
    this.sessions    = new SessionsResource(this);
    this.experiments = new ExperimentsResource(this);
    this.branding    = new BrandingResource(this);
    this.usage       = new UsageResource(this);
  }

  /** Set token after login */
  setToken(token) {
    this.token = token;
  }

  /** Internal HTTP request */
  async _request(method, path, { body, params } = {}) {
    const url    = new URL(this.baseUrl + path);
    if (params) {
      Object.entries(params).forEach(([k, v]) => v != null && url.searchParams.set(k, v));
    }

    const headers = { 'Content-Type': 'application/json' };
    if (this.token) headers['Authorization'] = `Bearer ${this.token}`;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const res = await fetch(url.toString(), {
        method,
        headers,
        body:   body ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });

      clearTimeout(timer);
      const data = await res.json().catch(() => ({}));

      if (!res.ok) {
        const err = new Error(data.error || `HTTP ${res.status}`);
        err.status  = res.status;
        err.code    = data.code;
        err.details = data.details;
        throw err;
      }
      return data;
    } catch (err) {
      clearTimeout(timer);
      throw err;
    }
  }

  _get(path, params)  { return this._request('GET',    path, { params }); }
  _post(path, body)   { return this._request('POST',   path, { body }); }
  _put(path, body)    { return this._request('PUT',    path, { body }); }
  _patch(path, body)  { return this._request('PATCH',  path, { body }); }
  _delete(path)       { return this._request('DELETE', path); }
}

// ─── Resources ────────────────────────────────────────────────────────────────

class AuthResource {
  constructor(sdk) { this._sdk = sdk; }

  /**
   * Login and automatically set the token on the SDK instance.
   * @param {string} email
   * @param {string} password
   * @param {string} [orgSlug]  Required for multi-tenant mode
   */
  async login(email, password, orgSlug) {
    const data = await this._sdk._post('/auth/login', { email, password, orgSlug });
    if (data.token) this._sdk.setToken(data.token);
    return data;
  }

  /** Register a new organization + owner account */
  registerOrg(orgName, orgSlug, email, password, name) {
    return this._sdk._post('/auth/register-org', { orgName, orgSlug, email, password, name });
  }

  /** Get current authenticated user */
  me() { return this._sdk._get('/auth/me'); }

  /** Logout current session */
  logout() { return this._sdk._post('/auth/logout'); }
}

class ClientsResource {
  constructor(sdk) { this._sdk = sdk; }

  /** List clients with optional filters */
  list({ status, search, page = 1, limit = 50 } = {}) {
    return this._sdk._get('/clients', { status, search, page, limit });
  }

  /** Get a single client by ID */
  get(id) { return this._sdk._get(`/clients/${id}`); }

  /** Create a new client */
  create(data) { return this._sdk._post('/clients', data); }

  /** Update a client */
  update(id, data) { return this._sdk._put(`/clients/${id}`, data); }

  /** Delete a client */
  delete(id) { return this._sdk._delete(`/clients/${id}`); }
}

class BroadcastsResource {
  constructor(sdk) { this._sdk = sdk; }

  list()       { return this._sdk._get('/broadcasts'); }
  get(id)      { return this._sdk._get(`/broadcasts/${id}`); }
  create(data) { return this._sdk._post('/broadcasts', data); }
  send(id)     { return this._sdk._post(`/broadcasts/${id}/send`); }
}

class PaymentsResource {
  constructor(sdk) { this._sdk = sdk; }

  list({ status } = {})   { return this._sdk._get('/payments', { status }); }
  get(id)                 { return this._sdk._get(`/payments/${id}`); }
  create(data)            { return this._sdk._post('/payments', data); }
  confirm(id)             { return this._sdk._post(`/payments/${id}/confirm`); }
}

class AnalyticsResource {
  constructor(sdk) { this._sdk = sdk; }

  dashboard()     { return this._sdk._get('/analytics'); }
  revenue()       { return this._sdk._get('/analytics/revenue'); }
  aiUsage()       { return this._sdk._get('/analytics/ai'); }
}

class AIResource {
  constructor(sdk) { this._sdk = sdk; }

  /** Send a message and get AI reply */
  reply(clientId, message) {
    return this._sdk._post('/ai/reply', { clientId, message });
  }

  /** Get AI reply suggestions */
  suggest(clientId, context) {
    return this._sdk._post('/ai/suggest', { clientId, context });
  }
}

class SessionsResource {
  constructor(sdk) { this._sdk = sdk; }

  list()               { return this._sdk._get('/sessions'); }
  revoke(sessionId)    { return this._sdk._delete(`/sessions/${sessionId}`); }
  revokeAll()          { return this._sdk._delete('/sessions'); }
}

class ExperimentsResource {
  constructor(sdk) { this._sdk = sdk; }

  list()                              { return this._sdk._get('/experiments'); }
  create(data)                        { return this._sdk._post('/experiments', data); }
  start(id)                           { return this._sdk._put(`/experiments/${id}/start`); }
  results(id)                         { return this._sdk._get(`/experiments/${id}/results`); }
  setWinner(id, winner)               { return this._sdk._post(`/experiments/${id}/winner`, { winner }); }
  recordConversion(id, clientId)      { return this._sdk._post(`/experiments/${id}/convert`, { clientId }); }
}

class BrandingResource {
  constructor(sdk) { this._sdk = sdk; }

  get()          { return this._sdk._get('/branding'); }
  update(data)   { return this._sdk._put('/branding', data); }
  initDomain(domain)  { return this._sdk._post('/domains/initiate', { domain }); }
  verifyDomain()      { return this._sdk._post('/domains/verify'); }
  getDomain()         { return this._sdk._get('/domains'); }
}

class UsageResource {
  constructor(sdk) { this._sdk = sdk; }

  monthly()     { return this._sdk._get('/usage/monthly'); }
  topFeatures() { return this._sdk._get('/usage/top-features'); }
}

// ─── Export ───────────────────────────────────────────────────────────────────

// CommonJS + ESM compatible
if (typeof module !== 'undefined' && module.exports) {
  module.exports = ClientFlowSDK;
  module.exports.default = ClientFlowSDK;
}
if (typeof window !== 'undefined') {
  window.ClientFlowSDK = ClientFlowSDK;
}

export default ClientFlowSDK;
